proposals/0002-gep.md

```diff
@@ -0,0 +1,140 @@
+---
+MEP: 0002
+Title: Unifed API
+Discussion: Link
+Implementation: Link
+---
+
+# TITLE: Unified API
+
+## Abstract
+
+This GEP discusses the unified API that handles the interaction with LLM providers. This represents the feature set for MVP.
+
+## Motivation
+
+
+The Unified API abstracts much of the overhead of interacting with LLM providers.
+It stores all the provider specific logic in a single place and allowing developers to focus on the core logic. It also introduces durability to LLM apps. Additionally, unifying the API allows optimizations to be leveraged such as latency and cost improvements.
+
+
+### Requirements
+
+- R1: User should be able to specify specific models or select auto mode (selection handled by Glide)
+- R2: Should be a Hertz Client
+- R3: Should take a user request as input
+- R4: Handles all provider specific logic
+- R5: Easily maintained
+- R6: API schemas must unify common LLM request params (e.g. topP, N, etc)
+- R7: API routes must unify common LLM endpoints/API
+- R8: API schemas should allow to customize provider-specific params (e.g. a different messages for OpenAI/Anthropic)
+
+
+## Design
+
+
+- A Hertz client that handles an incoming request and returns a reponse from that provider.
+- Provider specific clients
+
+### Routes
+```yaml
+routes:
+    chat: /chat/{pool-id}
```

Comment:
It would be more trivial to have API in this format:

```
/v1/pools/{pool-id}/chat/
/v1/pools/{pool-id}/complete/
...
```

Reply:
Thinking about these routes I would call `complete` API `completions` as REST respects nouns:
```
/v1/pools/{pool-id}/completions/
```